benchmarks/bench_tsne_mnist.py

```diff
@@ -0,0 +1,178 @@
+
+
+memory = Memory(os.path.join(LOG_DIR, 'mnist_tsne_benchmark_data'),
+                mmap_mode='r')
+
+
+@memory.cache
+def load_data(dtype=np.float32, order='C', shuffle=True, seed=0):
+    """Load the data, then cache and memmap the train/test split"""
+    print("Loading dataset...")
+    data = fetch_mldata('MNIST original')
+
+    X = check_array(data['data'], dtype=dtype, order=order)
+    y = data["target"]
+
+    if shuffle:
+        X, y = _shuffle(X, y, random_state=seed)
+
+    # Normalize features
+    X /= 255
+    return X, y
+
+
+def nn_accuracy(X, X_embedded, k=1):
+    """Accuracy of the first nearest neighbor"""
+    knn = NearestNeighbors(n_neighbors=1, n_jobs=-1)
+    _, neighbors_X = knn.fit(X).kneighbors()
+    _, neighbors_X_embedded = knn.fit(X_embedded).kneighbors()
+    return np.mean(neighbors_X == neighbors_X_embedded)
+
+
+def tsne_fit_transform(model, data):
+    transformed = model.fit_transform(data)
+    return transformed, model.n_iter_
+
+
+def sanitize(filename):
+    return filename.replace("/", '-').replace(" ", "_")
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser('Benchmark for t-SNE')
+    parser.add_argument('--order', type=str, default='C',
+                        help='Order of the input data')
+    parser.add_argument('--perplexity', type=float, default=30)
+    parser.add_argument('--bhtsne', action='store_true',
+                        help="if set and the reference bhtsne code is "
+                        "correctly installed, run it in the benchmark.")
+    parser.add_argument('--all', action='store_true',
+                        help="if set, run the benchmark with the whole MNIST."
+                             "dataset. Note that it will take up to 1 hour.")
+    parser.add_argument('--profile', action='store_true',
+                        help="if set, run the benchmark with a memory "
+                             "profiler.")
+    parser.add_argument('--verbose', type=int, default=0)
+    parser.add_argument('--n_jobs', type=int, nargs="+", default=2,
+                        help="Number of CPU used to fit sklearn.TSNE")
+    parser.add_argument('--pca-components', type=int, default=50,
+                        help="Number of principal components for "
+                             "preprocessing.")
+    args = parser.parse_args()
+
+    X, y = load_data(order=args.order)
+
+    if args.pca_components > 0:
+        t0 = time()
+        X = PCA(n_components=args.pca_components).fit_transform(X)
+        print("PCA preprocessing down to {} dimensions took {:0.3f}s"
+              .format(args.pca_components, time() - t0))
+
+    methods = []
+
+    # Put TSNE in methods
+    if isinstance(args.n_jobs, int):
+        tsne = TSNE(n_components=2, init='pca', perplexity=args.perplexity,
+                    verbose=args.verbose, n_jobs=args.n_jobs, n_iter=1000)
+        methods += [("sklearn TSNE",
+                     lambda data: tsne_fit_transform(tsne, data))]
```

Comment:
Complete nitpick, but I prefer to use the '.append' method, which is more explicite